targets/wasm_exec.js

```diff
@@ -298,11 +298,22 @@
 
 					// func finalizeRef(v ref)
 					"syscall/js.finalizeRef": (v_ref) => {
-						// Note: TinyGo does not support finalizers so this should never be
-						// called.
-						console.error('syscall/js.finalizeRef not implemented');
+						// Note: TinyGo does not support finalizers so this is only called
+						// for one specific case, by js.go:jsString. and can/might leak memory.
+						const id = mem().getUint32(unboxValue(v_ref), true);
+						if (this._goRefCounts?.[id] !== undefined) {
+							console.log("syscall/js.finalizeRef", id, this._goRefCounts[id]);
```

Comment:
Why this extra check?
1. `this._goRefCounts` won't be undefined, so the `?.` operator is not needed here.
2. I don't see an extra check in wasm_exec.js from upstream Go, so why would it be needed here? (I think it can be removed, or if it can't that indicates there's probably a bug somewhere else).

Also, can you remove these `console.log` logging statements? All they will do is spam the console without providing useful information to the developer (and if the developer actually wants to debug the finalizer, they can easily insert some temporary debug loggin).

Reply:
it actually is needed as in it doesn't crash but it's also not set - index seems to always be 0 and the ref not exist in my case so it's defensive

Reply:
@aykevl I updated the description, PTAL. If you have a (manual or otherwise) test to suggest where it wouldn't be unset/0 lmk, I would feel better seeing the code for `--` exercised at least once

Reply:
> I don't see an extra check in wasm_exec.js from upstream Go, so why would it be needed here?

I suppose that means this is either also a bug in big Go or else as @aykevl said a bug someplace else?

Reply:
> I suppose that means this is either also a bug in big Go

No, big go "works": it doesn't get invalid ids/not found

<img width="351" alt="Screenshot 2024-07-23 at 7 55 13 AM" src="https://github.com/user-attachments/assets/fc25070c-fd07-498e-9d33-6c8d32b66fca">

(added the same if/else logging there)

```javascript
					// func finalizeRef(v ref)
					"syscall/js.finalizeRef": (sp) => {
						sp >>>= 0;
						const id = this.mem.getUint32(sp + 8, true);
						if (this._goRefCounts?.[id] !== undefined) {
							console.log("syscall/js.finalizeRef", id, this._goRefCounts[id]);
							this._goRefCounts[id]--;
							if (this._goRefCounts[id] === 0) {
								const v = this._values[id];
								this._values[id] = null;
								this._ids.delete(v);
								this._idPool.push(id);
							}
						} else {
							console.log("syscall/js.finalizeRef: unknown id", id);
						}
					},
```

So yes there is a leak or issue, thus the comments and the log, but at least it's functional instead of making every single wasm js non functional like it is today afaik

Reply:
to see (modified) big go's version https://github.com/grol-io/grol/blob/big_go_wasm/wasm/wasm_exec.js#L302-L303